### DISCUSSION TITLE
Bug or feature? sa.delete causes exception when stream_results = True

### DISCUSSION THREAD

**Discussion by Thf772** [sqlalchemy/sqlalchemy#8255] [Usage Questions]
Hello, I have an error in some code I'm working on, but I am not sure whether it's from incorrect usage on my part or from an actual SQLAlchemy issue.

When I use an Engine initialized with `stream_results` set to `True`, I cannot use `sa.delete` statements (and possibly other DML statements). It seems SQLAlchemy is trying to prefix the generated SQL with the code handling streamed results for SELECT statements.

Tested with:
 - `SQLAlchemy==1.4.39`
 - `psycopg2==2.9.3`
 - PostgreSQL 14.4 on Ubuntu 20.04 AARCH64
 - Python 3.10.4 on Windows 10

### To reproduce:

```python
import sqlalchemy as sa

engine = sa.create_engine(
    connection_string,
    future = True,
    execution_options = {
        "stream_results" : True,
    },
)
Session = sa.orm.sessionmaker(
    engine,
    autoflush = False,
)

# Define or import ORM model
from . import dbmodel

with Session() as sess:
    sess.execute(sa.delete(dbmodel.MyModelClass))
```

### Actual result

The above code results in an exception. **It does not when removing `"stream_results" : True`.**

<details><summary>Stacktrace</summary>

```text
# Excluded non-SQLAlchemy stacks

  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\orm\session.py", line 1712, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\base.py", line 1631, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\sql\elements.py", line 332, in _execute_on_connection
    return connection._execute_clauseelement(
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\base.py", line 1498, in _execute_clauseelement
    ret = self._execute_context(
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\base.py", line 1862, in _execute_context
    self._handle_dbapi_exception(
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\base.py", line 2043, in _handle_dbapi_exception
    util.raise_(
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\util\compat.py", line 208, in raise_
    raise exception
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\base.py", line 1819, in _execute_context
    self.dialect.do_execute(
  File "C:\Program Files\Python310\lib\site-packages\sqlalchemy\engine\default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near "DELETE"
LINE 1: DECLARE "c_20b507e3430_9" CURSOR WITHOUT HOLD FOR DELETE FRO...
                                                          ^

[SQL: DELETE FROM my_model_table]
(Background on this error at: https://sqlalche.me/e/14/f405)
```
</details>

Even with `stream_results` on, issuing a `print(sa.delete(dbmodel.MyModelClass)` yields the correct SQL (without `DECLARE CURSOR`) but using `sess.execute(str(sa.delete(dbmodel.MyModelClass)))` still produces the exception. It looks like the session or engine is trying to add the `DECLARE CURSOR` snippet regardless of the statement type.

### Questions

Is my usage of `sa.delete` the proper way to wipe an ORM table *without loading its contents*?  
Would I avoid this exception by issuing the delete in another way?   
Should I open a proper issue (aka is this an actual bug)?

I should add that I am using SQLAlchemy as a workspace backend, to store data without saturating the RAM. the table I am trying to delete can contain tens of thousands of rows that are entirely re-created from time to time. This is why it is impossible to load them before deleting them.

Thanks for your help.

**Comment by zzzeek** [comment]
it's not safe at all to put "stream_results" in the engine-level execution options as server side cursors are implied by it and are not supported at all with many kinds of statements including DELETE.    there is a deprecated create_engine() option called server_side_cursors = True which is understood by the psycopg2 dialect that will selectively use a server side cursor, but this doesnt help "memory" so much when using the ORM as the ORM buffers all rows when SELECTing in any case unless "stream results" is used (it loads no rows at all when using delete()).

> I should add that I am using SQLAlchemy as a workspace backend, to store data without saturating the RAM. the table I am trying to delete can contain tens of thousands of rows that are entirely re-created from time to time. 

right so just use `delete()` as you are doing but don't try to set a global server side cursor behavior as that's not supported by the psycopg2 driver.
